bed/modify.go

```diff
@@ -92,3 +92,178 @@ func MergeHighMem(records []Bed, mergeAdjacent bool) []Bed {
+// and should be removed.
+func removeBedsWithNoTerritory(records []Bed) ([]Bed, bool) {
+	var answer []Bed
+	var threeDmidpoint int
+	var violation = false
+	for i := 1; i < len(records); i++ {
+		if records[i-1].Chrom == records[i].Chrom {
+			threeDmidpoint = (records[i-1].ChromEnd - records[i-1].Score + records[i].ChromStart + records[i].Score) / 2
+			if threeDmidpoint < records[i-1].ChromEnd {
+				violation = true
+			} else {
+				answer = append(answer, records[i-1])
+			}
+		} else {
+			answer = append(answer, records[i-1])
+		}
+	}
+	answer = append(answer, records[len(records)-1])
+
+	return answer, violation
+}
+
+//runUntilNoNewHidden directs our mergeKeepLowNameAndScore function to keep updating until no changes occur
+func runUntilNoNewHidden(records []Bed) []Bed {
+	var newHidden bool
+	records, newHidden = mergeKeepLowNameAndScore(records)
+	if newHidden {
+		runUntilNoNewHidden(records)
+	}
+	return records
+}
+
+// FillSpaceHiddenValue accepts a slice of Bed structs and a reference genome as a map[string]chromInfo.ChromInfo and returns a
+// slice of Bed structs that assigns each genomic position to the nearest feature in 3D space from the input bed, using
+// the input bed scores to represent "hidden values", or the distance from that position to its nearest TSS in 3D space.
+func FillSpaceHiddenValue(records []Bed, genome map[string]chromInfo.ChromInfo) []Bed {
+	records = runUntilNoNewHidden(records)
+	records = removeRecordsOnEmptyChrom(records, genome)
+	var violation = true
+	for violation {
+		records, violation = removeBedsWithNoTerritory(records)
+	}
+	var answer = make([]Bed, 0)
+	var threeDMidpoint int
+	var currAnswer = Bed{Chrom: records[0].Chrom, ChromStart: 0, ChromEnd: records[0].ChromEnd, Name: records[0].Name, FieldsInitialized: 4}
+
+	for i := 1; i < len(records); i++ {
+		if records[i].Chrom != currAnswer.Chrom {
+			currAnswer.ChromEnd = genome[records[i-1].Chrom].Size
+			if currAnswer.ChromEnd < currAnswer.ChromStart {
+				log.Print(records[i-1])
+				log.Print(currAnswer)
+				log.Fatalf("Died on new chrom.")
+			}
+			answer = append(answer, currAnswer)
+			currAnswer.Chrom = records[i].Chrom
+			currAnswer.ChromStart = 0
+		} else if currAnswer.Name == records[i].Name && currAnswer.Chrom == records[i].Chrom {
+			currAnswer.ChromStart = numbers.Min(currAnswer.ChromStart, records[i].ChromStart)
+			currAnswer.ChromEnd = numbers.Max(currAnswer.ChromEnd, records[i].ChromEnd)
+			currAnswer.Score = numbers.Min(currAnswer.Score, records[i].Score)
+		} else {
+			threeDMidpoint = (currAnswer.ChromEnd - records[i-1].Score + records[i].ChromStart + records[i].Score) / 2
+			currAnswer.ChromEnd = threeDMidpoint + 1
+			currAnswer.Name = records[i-1].Name
+			if currAnswer.ChromEnd-currAnswer.ChromStart < 0 {
+				log.Fatalf("Died in loop.")
+			}
+			answer = append(answer, currAnswer)
+			currAnswer.ChromStart = threeDMidpoint + 1
```

Comment:
Note to self: note how currAnswer.ChromStart is set to midpoint (after writing out previous currAnswer entry to output file) so that the next currAnswer entry starts at the midpoint (and can end at the TSS of the next gene, or a calculated 3d midpoint for another distal gene based on the closest wormhole, but both cases are covered by the "else" aka threeDMidpoint case)